x/sync/network_server_test.go

```diff
@@ -0,0 +1,287 @@
+			require.NoError(err)
+			require.LessOrEqual(len(bytes), int(test.request.BytesLimit))
+		})
+	}
+}
+
+func Test_Server_GetChangeProof(t *testing.T) {
+	require := require.New(t)
+
+	r := rand.New(rand.NewSource(1)) // #nosec G404
+	trieDB, _, err := generateTrieWithMinKeyLen(t, r, defaultRequestKeyLimit, 1)
+	require.NoError(err)
+
+	startRoot, err := trieDB.GetMerkleRoot(context.Background())
+	require.NoError(err)
+
+	// create changes
+	for x := 0; x < 600; x++ {
+		view, err := trieDB.NewView()
+		require.NoError(err)
+
+		key := make([]byte, r.Intn(100))
+		_, err = r.Read(key)
+		require.NoError(err)
+
+		val := make([]byte, r.Intn(100))
+		_, err = r.Read(val)
+		require.NoError(err)
+
+		err = view.Insert(context.Background(), key, val)
+		require.NoError(err)
+
+		deleteKeyStart := make([]byte, r.Intn(10))
+		_, err = r.Read(deleteKeyStart)
+		require.NoError(err)
+
+		it := trieDB.NewIteratorWithStart(deleteKeyStart)
+		if it.Next() {
+			err = view.Remove(context.Background(), it.Key())
+			require.NoError(err)
+		}
+		require.NoError(it.Error())
+		it.Release()
+
+		require.NoError(view.CommitToDB(context.Background()))
+	}
+
+	endRoot, err := trieDB.GetMerkleRoot(context.Background())
+	require.NoError(err)
+
+	tests := map[string]struct {
+		request             *ChangeProofRequest
+		expectedErr         error
+		expectedResponseLen int
+		nodeID              ids.NodeID
+		proofNil            bool
+	}{
+		"proof too small": {
+			request: &ChangeProofRequest{
+				StartingRoot: startRoot,
+				EndingRoot:   endRoot,
+				KeyLimit:     defaultRequestKeyLimit,
+				BytesLimit:   1000,
+			},
+			proofNil: true,
+		},
+		"byteslimit is 0": {
+			request: &ChangeProofRequest{
+				StartingRoot: startRoot,
+				EndingRoot:   endRoot,
+				KeyLimit:     defaultRequestKeyLimit,
+				BytesLimit:   0,
+			},
+			proofNil: true,
+		},
+		"keylimit is 0": {
+			request: &ChangeProofRequest{
+				StartingRoot: startRoot,
+				EndingRoot:   endRoot,
+				KeyLimit:     defaultRequestKeyLimit,
+				BytesLimit:   0,
+			},
+			proofNil: true,
+		},
+		"keys out of order": {
+			request: &ChangeProofRequest{
+				StartingRoot: startRoot,
+				EndingRoot:   endRoot,
+				KeyLimit:     defaultRequestKeyLimit,
+				BytesLimit:   constants.DefaultMaxMessageSize,
+				Start:        []byte{1},
+				End:          []byte{0},
+			},
+			proofNil: true,
+		},
+		"key limit too large": {
+			request: &ChangeProofRequest{
+				StartingRoot: startRoot,
+				EndingRoot:   endRoot,
+				KeyLimit:     2 * defaultRequestKeyLimit,
+				BytesLimit:   constants.DefaultMaxMessageSize,
+			},
+			expectedResponseLen: defaultRequestKeyLimit,
+		},
+		"bytes limit too large": {
+			request: &ChangeProofRequest{
+				StartingRoot: startRoot,
+				EndingRoot:   endRoot,
+				KeyLimit:     defaultRequestKeyLimit,
+				BytesLimit:   2 * constants.DefaultMaxMessageSize,
+			},
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
```

Comment:
We should make a new `require := require.New(t)` in each subtest so that if it errors, it'll include the name of the test that failed.